libandroid-navigation-ui/src/test/java/com/mapbox/services/android/navigation/ui/v5/route/MapRouteLineTest.java

```diff
@@ -42,111 +45,235 @@
+    CountDownLatch latchRunnablePrimary = new CountDownLatch(1);
+    CountDownLatch latch = new CountDownLatch(1);
+    Handler handlerFeatures = mock(Handler.class);
+    buildFeatureProcessingTask(routes, routeLine, handlerFeatures);
+    Handler handlerPrimary = mock(Handler.class);
+    buildPrimaryRouteUpdateTask(routeLine, handlerPrimary);
     routeLine.draw(routes);
+    latchRunnableFeatures.await(500, TimeUnit.MILLISECONDS);
+    verify(handlerFeatures).post(runnableFeatures.capture());
+    runnableFeatures.getValue().run();
+    latchRunnablePrimary.await(500, TimeUnit.MILLISECONDS);
+    verify(handlerPrimary).post(runnablePrimary.capture());
+    runnablePrimary.getValue().run();
+    ArgumentCaptor<Runnable> runnableUpdatePrimary = ArgumentCaptor.forClass(Runnable.class);
+    Handler handlerUpdatePrimary = mock(Handler.class);
+    buildPrimaryRouteUpdateTask(routeLine, handlerUpdatePrimary);
 
     routeLine.updatePrimaryRouteIndex(1);
+    latchRunnablePrimary.await(500, TimeUnit.MILLISECONDS);
+    verify(handlerUpdatePrimary).post(runnableUpdatePrimary.capture());
+    runnableUpdatePrimary.getValue().run();
 
+    latch.await(500, TimeUnit.MILLISECONDS);
     verify(routeLineSource, times(4)).setGeoJson(any(FeatureCollection.class));
   }
 
   @Test
-  public void updatePrimaryIndex_newPrimaryRouteIndexUpdated() throws IOException {
-    GeoJsonSource mockedRouteLineSource = mock(GeoJsonSource.class);
-    GeoJsonSource mockedWayPointSource = mock(GeoJsonSource.class);
-    List<Layer> anyRouteLayers = buildMockLayers();
+  public void updatePrimaryIndex_newPrimaryRouteIndexUpdated() throws Exception {
+    GeoJsonSource routeLineSource = mock(GeoJsonSource.class);
+    GeoJsonSource wayPointSource = mock(GeoJsonSource.class);
+    List<Layer> routeLayers = buildMockLayers();
+    MapRouteLine routeLine = new MapRouteLine(routeLineSource, wayPointSource, routeLayers);
     List<DirectionsRoute> routes = new ArrayList<>();
     routes.add(buildTestDirectionsRoute());
     routes.add(buildTestDirectionsRoute());
     routes.add(buildTestDirectionsRoute());
     routes.add(buildTestDirectionsRoute());
-    MapRouteLine routeLine = new MapRouteLine(mockedRouteLineSource, mockedWayPointSource, anyRouteLayers);
+    ArgumentCaptor<Runnable> runnableFeatures = ArgumentCaptor.forClass(Runnable.class);
+    ArgumentCaptor<Runnable> runnablePrimary = ArgumentCaptor.forClass(Runnable.class);
+    CountDownLatch latchRunnableFeatures = new CountDownLatch(1);
+    CountDownLatch latchRunnablePrimary = new CountDownLatch(1);
+    CountDownLatch latch = new CountDownLatch(1);
+    Handler handlerFeatures = mock(Handler.class);
+    buildFeatureProcessingTask(routes, routeLine, handlerFeatures);
+    Handler handlerPrimary = mock(Handler.class);
+    buildPrimaryRouteUpdateTask(routeLine, handlerPrimary);
     routeLine.draw(routes);
+    latchRunnableFeatures.await(500, TimeUnit.MILLISECONDS);
```

Comment:
do we really need 500 Millisecond of await? it wastes Test-time

Reply:
You're right. Tried with `CountDownLatch#countDown` but tests were failing / flaky. As explained above we can implement the tasks as `Runnable`s and create the thread in the call-site. That way we'll be able to execute the `run` synchronously. In any case, we can revisit later, in a future PR.